### ISSUE TITLE
JAX upstream: `betaln(a, b)` is wrong by `min(a, b)` when `min(a, b) / max(a, b)` is subnormal

### ISSUE THREAD

**Issue by nstarman** [JAXtronomy/spexial#27]
Found while testing `spexial` against JAX. Tracked here so it can be opened on [jax-ml/jax](https://github.com/jax-ml/jax/issues) later; everything below is written to be submitted as-is, and includes a one-line fix verified to give 0 ulp on every affected case while leaving the rest bit-identical.

**What it means for `spexial`:** this is the only one of the four we have to work around. `comb` evaluates large `N` through `1 / ((N + 1) B(N - k + 1, k + 1))`, and the flush made `comb(N, 1)` a factor of `e**-2` — 86% — low for every `N` above `2**1023`. Fixed in 2048c5d by evaluating the Beta term from its asymptotic form where the quotient would be subnormal, which is exactly where that asymptotic is itself exact. Written as a formula rather than a special case so that it still differentiates in both arguments.

---
### Description

`jax.scipy.special.betaln` is exact across most of its range, but returns an answer too large by exactly `min(a, b)` once the ratio of its arguments falls below the smallest normal double.

For `betaln(2, N)` the cutoff is exactly `N = 2**1023`, to the last bit: that is where `2 / N` stops being a normal double. `scipy.special.betaln` is correct throughout.

The consequence downstream is silent and large. `1 / ((N + 1) * B(N - k + 1, k + 1))` is the standard cancellation-free way to evaluate a binomial coefficient, and it comes out a factor of `e**-2` — 86% — low for every `N` above `2**1023`.

### Reproducing code

```python
import jax

jax.config.update("jax_enable_x64", True)
import numpy as np
import scipy.special as ss
from jax.scipy.special import betaln

tiny = np.finfo(np.float64).tiny
cases = [
    (1.0, 1e308),
    (2.0, 1e308),
    (1.5, 1e308),
    (2.0, 8.99e307),
    (2.0, 8.9e307),
    (3.0, 1e308),
    (2.0, 1e300),
]

print(
    f"{'a':>5} {'b':>12} {'a/b':>12} {'subnormal':>10} {'jax':>16} {'scipy':>16} {'err':>7}"
)
for a, b in cases:
    h = a / b
    j = float(betaln(a, b))
    s = float(ss.betaln(a, b))
    print(
        f"{a:>5} {b:>12.3e} {h:>12.4e} {str(0 < h < tiny):>10} {j:>16.8f} {s:>16.8f} {j - s:>+7.2f}"
    )
```

### Output

```
    a            b          a/b  subnormal              jax            scipy     err
  1.0   1.000e+308  1.0000e-308       True    -708.19620864    -709.19620864   +1.00
  2.0   1.000e+308  2.0000e-308       True   -1416.39241728   -1418.39241728   +2.00
  1.5   1.000e+308  1.5000e-308       True   -1062.41509520   -1063.91509520   +1.50
  2.0   8.990e+307  2.2247e-308       True   -1416.17947280   -1418.17947280   +2.00
  2.0   8.900e+307  2.2472e-308      False   -1418.15934965   -1418.15934965   +0.00
  3.0   1.000e+308  3.0000e-308      False   -2126.89547875   -2126.89547875   +0.00
  2.0   1.000e+300  2.0000e-300      False   -1381.55105580   -1381.55105580   +0.00
```

The error is exactly `a` whenever `a / b` is subnormal, and exactly zero otherwise. Verified against `mpmath` at 400 digits (`loggamma(1e308)` is about `7e310`, so the ~1400 being measured needs more than 311 digits of working precision — a lower `dps` silently reports zero error).

### Cause

`jax/_src/third_party/scipy/betaln.py`, in `algdiv`:

```
h = a / b
d = b + (a - 0.5)
...
u = d * lax.log1p(a / b)
```

XLA flushes subnormals to zero on CPU, so once `a / b` is subnormal the quotient becomes `0`, `log1p(0)` is `0`, and `u` — whose true value is `d * (a/b) ≈ b * (a/b) = a` — is dropped entirely. Hence an error of exactly `a`.

The Fortran `algdiv` this is derived from does not flush, which is why `scipy.special.betaln` is unaffected.

### Suggested fix

Where `h` has flushed to zero, `log1p(h)` would have been `h` to the last bit anyway, so `u = d * h = a * (d / b)` — and `d / b` is close to 1, so it cannot underflow. That makes the fix one line, and it changes nothing outside the flushed region because it is only reached when `h == 0`:

```python
u = jnp.where(h == 0, a * (d / b), d * lax.log1p(h))
```

Re-running the table above with that change gives `0.00e+00` error against mpmath for all four subnormal rows, and leaves every other row bit-identical.

### Unrelated observation, from the same measurements

`betaln` is also noticeably less accurate than SciPy's just above the `b >= 8` branch cut — `betaln(2, 8)` is `1.3e-06` off and `betaln(8, 20)` is `8.5e-08` off, where SciPy is exact. That is a separate matter from the flush above and is only mentioned here because the same script measured it.

---

*Verified with jax 0.11.1, jaxlib 0.11.1, CPU backend, x64 enabled.*

**Comment by nstarman** [comment]
Reported upstream: https://github.com/jax-ml/jax/issues/40653

Filed as a new issue. No duplicate — the open `betaln` issues upstream ([#39106](https://github.com/jax-ml/jax/issues/39106), [#38243](https://github.com/jax-ml/jax/issues/38243), [#34353](https://github.com/jax-ml/jax/issues/34353)) are about infinities, negative integers and the Hessian. Worth knowing: [#13267](https://github.com/jax-ml/jax/issues/13267) ("betaln is wildly inaccurate for large values", closed) is what introduced the `algdiv` path, so this is a bug in the fix for that one — I cited it upstream as context.

## Re-verified on jax 0.11.1 / jaxlib 0.11.1 / scipy 1.18.1

This issue is accurate throughout and the reproducer went up essentially as written. Every row of the table reproduces, the error is exactly `a` on flushed inputs and exactly zero otherwise, and the `algdiv` source matches the quoted lines.

**The one-line fix does everything claimed.** Tested against mpmath at 400 dps:

- flushed cases all land at **0 ulp** — `(1, 1e308)`, `(2, 1e308)`, `(1.5, 1e308)`, `(2, 8.99e307)`, `(0.5, 1.7e308)` bit-exact afterwards; for `(1.5, 1e308)` that is one ulp *better* than SciPy, which is off by -2.27e-13
- 500 random non-flushed pairs over `a ∈ (1e-3, 50)`, `b ∈ (8, 1e300)` come back **bit-identical** to current output

I added one implementation note upstream that isn't in this issue: the line today reads `d * lax.log1p(a / b)`, recomputing the quotient rather than using `h`. The patch has to switch it to `h` as well as add the `where`, or the guard tests one value and the branch uses another.

I also confirmed the flush directly rather than leaving it inferred: `float(1.0 / jnp.float64(1e308))` is `0.0` where numpy gives `1e-308`.

## One sharpening

> the cutoff is exactly `N = 2**1023`, to the last bit

One ULP off. `2**1023` itself is still correct — `2 / 2**1023` is the smallest *normal*, not subnormal. The first wrong result is the next float up:

```
8.988465674311579e+307  (2**1023 - 1ulp)   normal      err +0.00
8.98846567431158e+307   (2**1023)          normal      err +0.00
8.988465674311582e+307  (2**1023 + 1ulp)   subnormal   err +2.00
```

Immaterial to the bug, but I stated it as "one ULP above 2**1023" upstream since the report leans on the threshold being exact.

## On the secondary observation

The `b >= 8` accuracy note is real and somewhat broader than described here — it is the whole `algdiv` branch near the cut, not just the two values sampled:

```
  2.0    7.9     -5.33e-15   <- lgamma branch
  1.0    8.0      9.30e-07   <- algdiv branch
  2.0    8.0      1.29e-06
  3.0    8.0      1.34e-06
  0.5    8.0      5.58e-07
  5.0    9.0      7.73e-07
  8.0   20.0      8.49e-08
  2.0  100.0      1.05e-10
```

About nine orders of magnitude lost crossing `b = 8`, recovering as `b` grows; SciPy is exact at all of these. I included it upstream as a clearly-marked separate observation with an offer to split it out.
